physical/raft/raft_autopilot.go

```diff
@@ -0,0 +1,710 @@
+		CleanupDeadServers:          false,
+		LastContactThreshold:        10 * time.Second,
+		LastContactFailureThreshold: 24 * time.Hour,
+		MaxTrailingLogs:             1000,
+		MinQuorum:                   3,
+		ServerStabilizationTime:     10 * time.Second,
+	}
+}
+
+func (b *RaftBackend) AutopilotHCLConfig() (*AutopilotConfig, error) {
+	defaultConfig := b.defaultAutopilotConfig()
+
+	config := b.conf["autopilot"]
+	if config == "" {
+		return nil, nil
+	}
+
+	var configs []*AutopilotConfig
+	err := jsonutil.DecodeJSON([]byte(config), &configs)
+	if err != nil {
+		return nil, errwrap.Wrapf("failed to decode autopilot config: {{err}}", err)
+	}
+	if len(configs) != 1 {
+		return nil, fmt.Errorf("expected a single block of autopilot config")
+	}
+
+	hclConfig := configs[0]
+
+	// Sanitize input
+	if hclConfig.LastContactThreshold == 0 {
+		hclConfig.LastContactThreshold = defaultConfig.LastContactThreshold
+	}
+	if hclConfig.LastContactFailureThreshold == 0 {
+		hclConfig.LastContactFailureThreshold = defaultConfig.LastContactFailureThreshold
+	}
+	if hclConfig.MaxTrailingLogs == 0 {
+		hclConfig.MaxTrailingLogs = defaultConfig.MaxTrailingLogs
+	}
+	if hclConfig.MinQuorum == 0 {
+		hclConfig.MinQuorum = defaultConfig.MinQuorum
+	}
+	if hclConfig.ServerStabilizationTime == 0 {
+		hclConfig.ServerStabilizationTime = defaultConfig.ServerStabilizationTime
+	}
+
+	return hclConfig, nil
+}
+
+func (b *RaftBackend) AutopilotDisabled() bool {
+	b.l.RLock()
+	disabled := b.disableAutopilot
+	b.l.RUnlock()
+	return disabled
+}
+
+// StartAutopilot puts autopilot subsystem to work. This should only be called
+// on the active node.
+func (b *RaftBackend) StartAutopilot(ctx context.Context) {
+	b.l.RLock()
+	defer b.l.RUnlock()
+
+	if b.autopilot == nil {
+		return
+	}
+
+	b.autopilot.Start(ctx)
+	b.followerHeartbeatTrackerStopCh = make(chan struct{})
+	go b.startFollowerHeartbeatTracker()
+}
+
+// AutopilotExecutionStatus represents the current status of the autopilot background go routines
+type AutopilotExecutionStatus string
+
+const (
+	AutopilotNotRunning   AutopilotExecutionStatus = "not-running"
+	AutopilotRunning      AutopilotExecutionStatus = "running"
+	AutopilotShuttingDown AutopilotExecutionStatus = "shutting-down"
+)
+
+func autopilotStatusToStatus(status autopilot.ExecutionStatus) AutopilotExecutionStatus {
+	switch status {
+	case autopilot.Running:
+		return AutopilotRunning
+	case autopilot.ShuttingDown:
+		return AutopilotShuttingDown
+	default:
+		return AutopilotNotRunning
+	}
+}
+
+func (b *RaftBackend) startFollowerHeartbeatTracker() {
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-b.followerHeartbeatTrackerStopCh:
+			return
+		case <-ticker.C:
+			b.l.RLock()
+			if b.autopilotConfig.CleanupDeadServers && b.autopilotConfig.LastContactFailureThreshold != 0 {
+				b.followerStates.l.RLock()
+				followers := b.followerStates.followers
+				b.followerStates.l.RUnlock()
```

Comment:
I don't think it's safe to release this lock before iterating over followers - what if someone calls Delete to remove one?

Reply:
Adjusted the locking. Let me know what you think.

Reply:
I meant the followerStates lock.

Reply:
Yep. That is the lock I adjusted. The lock is now released after iterating.

Reply:
What I also did was to grab the write lock only if there is a criteria match for marking the node as dead. Coz otherwise, we would hold the write lock far too often.

Reply:
```
                                b.followerStates.l.RLock()
				followers := b.followerStates.followers
				b.followerStates.l.RUnlock()
                                for id, state := range followers {
```

I worry that `range followers` is unsafe while we don't have `b.followerStates.l` locked, since followers is just a pointer to the same map, and FollowerStates.Delete could delete it from it at any time.

Reply:
Right. That's what I thought you suggested and have the lock held for the entire duration of range now. The diff here is pointing to the old code. Can you take a look at the latest changes and let me know if that addresses the concern?

Reply:
It's still not showing in github for some reason, but I've pulled the changes locally and see what you mean.  Thanks!